### PR TITLE
Allow builds with -std=c99 on LTS branch 2.7

### PR DESCRIPTION
**Pull request by gufe44**
Includes (partial) backports of #1198, #3190, #3421, #3422 and #3571.

### CODE REVIEW

File: tests/suites/main_test.function [hunk 1]
```diff
@@ -400,14 +400,21 @@ int main(int argc, const char *argv[])
...
-                                    strdup( params[i] );
+                                    malloc( len + 1 );
                                 if( unmet_dependencies[unmet_dep_count] == NULL )
                                 {
                                     mbedtls_fprintf( stderr,
                                                      "FATAL: Out of memory\n" );
                                     mbedtls_exit( MBEDTLS_EXIT_FAILURE );
                                 }
+                                for( j = 0; j < len; j++ )
+                                {
```
Comment (daverodgman) [inline tests/suites/main_test.function:413]: It would be simpler to call strncpy here

**Review by daverodgman — CHANGES_REQUESTED** [review]

**Review by daverodgman — APPROVED** [review]

File: library/entropy_poll.c [hunk 2]
```diff
@@ -44,6 +44,11 @@
  *  **********
  */
 
+#if defined(__linux__)
+/* Ensure that syscall() is available even when compiling with -std=c99 */
+#define _GNU_SOURCE
```
Comment (gilles-peskine-arm) [inline library/entropy_poll.c:49]: As reported in https://github.com/ARMmbed/mbedtls/issues/3432, `_GNU_SOURCE` should not be redefined if it's already defined.

This also applies to `_POSIX_C_SOURCE`, `_XOPEN_SOURCE` and any similar macro.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The changes are ok, except that the `_xxx_SOURCE` macros should only be defined if they aren't already defined.

Please add `-std=c99 -pedantic` to [`test_build_opt` in `all.sh`](https://github.com/ARMmbed/mbedtls/blob/v2.24.0/tests/scripts/all.sh#L1673), as is already the case in `mbedtls-2.16` and `development`, to avoid future regressions.

**Comment by gufe44** [comment]
Unfortunately, for personal reasons I won't be able to continue to work on the PR. You can close the PR now and resurrect the branch at any time.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
As with https://github.com/ARMmbed/mbedtls/pull/3658#pullrequestreview-528987682, I don't think these changes are correct, because they require a more recent POSIX version than what the long-time-support branch originally required.

**Comment by chris-jones-arm** [comment]
I agree with the above. This is being closed as these changes would not be compatible with older systems on the LTS.

**Pull request closed without merging**